Open Source as a Service (#1)
stefwalter/blog: content/technical/open-source-as-a-service.md

Title: Your Service is not Open Source
Date: 2020-02-07 18:35
Tags: open source, technical

> Slug: open-source-as-a-service

Reviewer: i think I prefer your "Open Source Services" term. It better communicates intent (open source as a service sounds like you are offering open source itself as a service, while our focus is on services here). It is also a nice play on OSS, but with a very nice & clear differentiation between Software vs. Service.

Reviewer 2: arent slugs just `hyphenate($title)` ?

Author: Thanks @gsaslis. I checked that name with some of the others in a call today, and everyone liked "Open Source Services" ... so good call. Will change the slug and start using that name elsewhere as well.

Category: Technical

> Summary: Sharing the source code to your SaaS is insufficient. We effectively enable users to contribute to the a running service for it to be Open Source.

Reviewer: How about 👇 ?

Open sourcing the code to your SaaS is insufficient. For a *service* to really be Open Source, we need to effectively enable users to contribute to the SaaS itself.

Author: Better. Thanks.

> Summary: Sharing the source code to your SaaS is insufficient. We effectively enable users to contribute to the a running service for it to be Open Source.

Reviewer 2: We effectively **need to** enable users to contribute to ~~the~~ a running service for it to be Open Source.

Author: Indeed. Thanks for catching that. I replaced this @gsaslis text.

> Sharing the source code to your SaaS is insufficient to make it

Reviewer: "Sharing the source code to" -> "Open Sourcing" ?

Author: Done.

actually be Open Source. Sounds self-contradictory?

Most services that espouse "Open Source", do so by simply
[throwing the code over the wall](https://github.com/travis-ci/travis-ci).
It's better than nothing, but really misses the point that powers
Open Source: enabling users to make a change to the software
they're using.

> Some other Open Source services, such as

Reviewer: I would change "Some other Open Source services" to save "Open Source Services" as the term to use only for our end game, not for what (gitlab|elastic) currently do. 

"Some other popular services powered by Open Source Software" 

(even though this might now be controversial in elastic's case...)

Reviewer 2: agreed. Coining a term is powerful

Author: Agree with coining the term. But I wouldn't like to apply the term to these projects (GitLab is Open Core + Managed Service and ElasticSearch has a dubious license now).

Author: Updated to use the text @gsaslis suggested in quotes.

[GitLab.com](https://about.gitlab.com/community/contribute/development/)
or [ElasticSearch](https://www.elastic.co/about/free-and-open) do include
the tools used to operate/deploy their service. Pause for applause 👏👏👏

But that's also insufficient to actually enable users to become
contributors effectively.

Much of the value of a service comes from things other than the code.
It comes from the infrastructure it runs on, the operational processes,
the monitoring, the backups, the metrics, the high availability, the

> scalability. It comes from the network effect, with other users, with other

Reviewer 3: ```suggestion
scalability, the data, the integration with other tools and services, and more. It comes from the network effect, with other users, with other
```

Author: I think integration with other tools and services is the network effect. So lets add that there.

interconnected services, from legal agreements, and so on.

Any non-toy Service is not reproducible.

Launching a clone of the service in order to effectively contribute
is a high barrier. Certain kinds of contributions clearly are possible.
But contributors are essentially asked to "fork" the above non-code
aspects of the service in order to iterate on their changes against
their real world use cases.

What's more, without true Open Source principles, we are missing a
way to have a community develop with a center of gravity around
the service itself. This leads to companies
[forking the service](https://aws.amazon.com/blogs/opensource/stepping-up-for-a-truly-open-source-elasticsearch)
in a way that typical open source projects with a true community
are resilient to.

> I believe that if we enable contributions to services, we reap

Reviewer: let's add 

"I believe that if we enable contributions to services, *rather than just the software*,"

Author: Clearer. Thanks.

the advantages of true Open Source. That means enabling users to make a
change to a running service and experience that change themselves.

You still with me?

Yes, that means opening a pull request against a running service

> and running that change before it's merged.

Reviewer 3: ```suggestion
and running that change before it's merged with no effort for the contributor.
```

Author: I'm not sure it's "no effort'. But I changed the second running to be "experiencing" to make it more clear.

to have one company or one individual invent everything from scratch.
So we work together across humanity to accomplish what no
individual team could. This is our legacy, and it has become
common place.

And yet there's a very simple principle that drives Open Source:

<div style="font-weight: bold; text-align: center;">
Open Source thrives when it converts some<br>small percentage of users into contributors.
</div>

If you interrupt that principle, you starve Open Source. Software as a
Service does just that: When someone else runs software for you, the
intuitive mechanisms for users to change that software are unavailable.

"But" you may say, "If the source code for that running service was **open**
then I could still change it."

Sure you could change the code in some component, but for any reasonably
complex Service you wouldn't be able to run your changes, experience your
changes against real world workflows, much less iterate on them until
they're good enough to share with others.

In reality the threat here is far more fundamental: Users of services
cannot readily change something in the service not only because the

> processes is operated by someone else ... but because such users of

Reviewer: ```suggestion
processes are operated by someone else ... but because such users of
```

a service have explicitly chosen not to be involved in operating
the software.

> If the primary way to use [Postgres was as a Service](https://aws.amazon.com/rds/),

Reviewer: "as a Service" ?

(i.e.  add quotes to clarify. This took me a few reads to ensure I understood the point)

Reviewer 2: Yes, how about:
"If the primary way how to use Postrgres, would have been to consume it [as a Service](), then that..."

Author: Thanks. Good catch. Updated to use quotes. It matches my flavor better ;)

Open Source ecosystem.

But recently, I've become convinced that if we enable an effective
contribution model on a service, one that doesn't require that the users
themselves become operators of the service ... then we can reconcile this
threat and Open Source will thrive on services instead of be starved by them.

# You want me to do what!?

To enable contributions by users of a service, users that have explicitly
chosen not to operate the software themselves, we have to set up a process
by which they can make a change to a running service and experience that
change themselves before others do.

For example, a user of such a hypothetical service should be able to:

 1. Discover which component of a service to contribute to.

 2. Make a non-sensical change (like add a printf style log statement) or change the spelling of a word.

 3. Experience that change when they use the service, or when it acts on their data.

After discussing this with others, I firmly believe we have all the techniques we need,
whether canary deployments, load balancing, continuous delivery,
infrastructure as code, and so on. It's been hard to find a single

> problem that an existing practice doesn't solve.

Reviewer 4: Not for your article (perhaps), but a thought: Achieving this would also mean that you have full control of the "real" deployment, and constantly exercise and improve its (re)deployment. The importance of that can hardly be over-estimated for a project's/company's long-term resilience.

> It turns out many engineering teams working have come to that same

Reviewer 2: s/many engineering teams working/many experienced engineering teams//

Author: Thanks.

conclusion. Rapid deployment of changes to services is a powerful
capability highly sought after. For example,
[GitHub deploys changes before they're merged](https://github.blog/2015-06-02-deploying-branches-to-github-com/).
Whereas
[Facebook rapidly deploys changes](https://engineering.fb.com/2017/08/31/web/rapid-release-at-massive-scale/)
to a few canary machines and scales each change up to production.
Everyone has their own version of continuous delivery. It's hard to
take an engineering team seriously if they haven't figured out how
their team members can get a change rapidly deployed.

As Open Source, we have all the ingredients to have a decisive

> advantage here. To effectively collaborate on services with

Reviewer: ```suggestion
advantage here. To effectively collaborate on  **Open Source Services**, 
```

contributors across humanity.

> Lets try to craft a playbook to achieve this basic capability

Reviewer 2: How about

"Lets try to craft a playbook to apply this basic capability that drives Open Source Software to Open Source Services: The ability..."

Using the newly coined term "OSServices"  twice for brainzzz

Author: Awesome. Added.

Title: Your Service is not Open Source

> Date: 2018-02-06 18:35

Reviewer 4: Do you have your time travel permit from the Ministry of Truth? :grin:

Author: Heh. Fixed.

Sharing the source code to your SaaS is insufficient to make it
actually be Open Source. Sounds self-contradictory?

Most services that espouse "Open Source", do so by simply
[throwing the code over the wall](https://github.com/travis-ci/travis-ci).
It's better than nothing, but really misses the point that powers
Open Source: enabling users to make a change to the software
they're using.

Some other Open Source services, such as
[GitLab.com](https://about.gitlab.com/community/contribute/development/)
or [ElasticSearch](https://www.elastic.co/about/free-and-open) do include
the tools used to operate/deploy their service. Pause for applause 👏👏👏

But thats is also insufficient to actually enable Open Source.

In order for a user of the service to begin contributing, they must
first deploy and operate their own clone: Something that they explicitly
chose not to do when they elected to use the software as a service.

Even if the user becomes an operator, and deploys a such clone, it
is devoid of their own data, devoid of other users, of connections
to other services, of infrastructure scalability, pipelines, monitoring,

> backups, ... you name it.

Reviewer 4: This is a bit harsh IMHO. First, the whole network of users, credentials, data etc. are the actual and precious assets of the "primary" SaaS, and exposing/exporting all of that is not only undesirable and inappropriate, it's outright illegal (data protection, private projects, exposing secrets, etc.). Also the whole point of hosting your own instance is that you can work with your own data, test it, and break it.

The second half (scalability, pipelines, monitoring, backups) just means that the deployment scripts are incomplete -- things like metrics collection or database replication should be part of the deployment; of course the "other" side of that (pagers, email etc.) is highly specific to a particular user/company. Also, for test instances you most likely don't want that extra overhead anyway.

I understand that this is primarily an argument for *not* hosting your own instance in this context, and for that it's valid -- but that's already handled in the first paragraph.

Author: Good points. I've tried to rewrite this and stick to objective claims. WDYT? Do you feel these paragraphs are more coherent now? I'd be happy for further critique, in areas where my subjectivity remains.

In any case, users are certainly not able to iterate on their changes
to a service against their real world use cases.

What's more, without true Open Source principles, we are missing a
way to have a community develop with a center of gravity around
the service itself. This leads to companies
[forking the service](https://aws.amazon.com/blogs/opensource/stepping-up-for-a-truly-open-source-elasticsearch)
in a way that typical open source projects with a true community
are resilient to.

I believe that if we enable contributions to services, we reap
the advantages of true Open Source. That means enabling users to make a
change to a running service and experience that change themselves.

You still with me?

Yes, that means opening a pull request against a running service

> and running that change before it's merged.

Reviewer 4: Connecting that to your point above -- the *only* way that this is feasible is if your own little shard of that SaaS with your PR operates on a cloned subset of the production data. If any random Joe Developer could submit a PR against gitlab that runs on the production data, what trust could you still place in the gitlab Saas? IMHO, it would be "none".

But again, I think having a playground with empty (or cloned data that belongs to you) is actually preferable, as it gives you the possibility of doing thorough, destructive, and risky tests.

Author: I'll cover this in follow up articles and playbook. I believe that your statements are just as true of team contributions.
Many teams only merge code once they have gotten it into production. See the GitHub internal contribution process.

But yes, I'm happy leaving this as a controvesial topic to be discussed further.

Reviewer: @martinpitt i think the crucial bit that is mentioned above is that this change would only affect your data / how only you consume the service. 

I think of it like this: 
many SaaS offer an option to try their "beta" features or become part of their "fast" channel, with latest changes not-yet fully tested on everyone else. Other SaaS A/B test features all the time, so it's not like all the users of the service experience the same service anyway. 

Switching to "contributor mode" for an Open Source Service would be very similar to those cases but the features you are beta testing are simply your own.

This is not as insane as it sounds.

It appears that every technique we need to enable such a capability
is already in use in modern deployment and operations methodology.
We just have to connect the dots.

# Why should I care?

I've been a part of Open Source for over 20 years now. Although I'm
just one contributor, together we've changed something fundamental
about the world. Our lives are too short, and software too complex,
to have one company or one individual invent everything from scratch.
So we work together across humanity to accomplish what no
individual team could. This is our legacy, and it has become

> common place.

Reviewer 4: I had to shed a tear of joy here :joy: Well written!

Author: Thanks!

 1. Discover which component of a service to contribute to.

 2. Make a non-sensical change (like add a printf style log statement) or change the spelling of a word.

 3. Experience that change when they use the service, or when it acts on their data.

After discussing this with others, I firmly believe we have all the techniques we need,
whether canary deployments, load balancing, continuous delivery,
infrastructure as code, and so on. It's been hard to find a single
problem that an existing practice doesn't solve.

It turns out many engineering teams working have come to that same
conclusion. Rapid deployment of changes to services is a powerful
capability highly sought after. For example,
[GitHub deploys changes before they're merged](https://github.blog/2015-06-02-deploying-branches-to-github-com/).
Whereas
[Facebook rapidly deploys changes](https://engineering.fb.com/2017/08/31/web/rapid-release-at-massive-scale/)
to a few canary machines and scales each change up to production.
Everyone has their own version of continuous delivery. It's hard to
take an engineering team seriously if they haven't figured out how
their team members can get a change rapidly deployed.

> As Open Source, I don't see us having pulled this off. Not yet.

Reviewer 4: This is something different, though, so IMHO this breaks the flow of argument here. This paragraph is about the engineers of the SaaS installation doing CD of their own code to their own official installation. I don't know how gitlab or e.g. the Testing Farm cluster handles that, but I can't imagine that CD is so rare in the FOSS world? (FWIW, this is exactly how we handle infra changes in our cockpit CI, even though that is not a public service).

Your plea above was that developers *outside* of that core developer team could submit a PR and get their own shard of the service (*not* get it deployed to the official primary SaaS). The latter is indeed much harder IMHO, but then again I am not aware of any non-open project which does that. In fact, by definition you couldn't even do something like "send a PR against github".

So, I agree that FOSS has not managed to set up this "get your own shard on a PR", and this is indeed something that would give FOSS infra projects a decisive advantage over their closed-source counterparts. But it's certainly not an aspect where FOSS infra projects lag behind closed-source ones. So this could be easy to msunderstand here.

Author: Indeed. This was sloppy of me. I removed a bunch of the "how do we do it" stuff for later posts because the article was getting too long. How does it look now?
